tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Grpc/GrpcLegacy/Client/CachedMetadataHelper.cs

```diff
@@ -0,0 +1,47 @@
+﻿// <copyright file="CachedMetadataHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Reflection;
+using System.Reflection.Emit;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Grpc.GrpcLegacy.Client
+{
+    internal static class CachedMetadataHelper<TMarkerType>
+    {
+        private static Func<object> _activator;
+
+        static CachedMetadataHelper()
+        {
+            CreateActivator(typeof(TMarkerType));
+        }
+
+        private static void CreateActivator(Type markerType)
+        {
+            var metadataType = markerType.Assembly.GetType("Grpc.Core.Metadata");
+
+            ConstructorInfo ctor = metadataType.GetConstructor(System.Type.EmptyTypes);
+
+            DynamicMethod createHeadersMethod = new DynamicMethod(
+                $"GrpcCoreCachedMetadataHelper",
+                metadataType,
+                null,
+                typeof(DuckType).Module,
+                true);
+
+            ILGenerator il = createHeadersMethod.GetILGenerator();
```

Comment:
I guess there probably is a reason, but why are you using generated IL here rather than just calling `Activator.CreateInstance(metadataType);` ?

Reply:
That's a good question. The short answer is "because that's what do elsewhere". There's a _relatively_ large difference in [the benchmark performance](https://andrewlock.net/benchmarking-4-reflection-methods-for-calling-a-constructor-in-dotnet/#the-results) between the methods, but also a significant startup cost. It may well be re-evaluating whether `Activator.CreateInstance()` is preferable globally, but I think it make sense to stay consistent with the other cases for now?